### PR TITLE
Optimize Crypto::Bcrypt performance

### PR DESCRIPTION
**Pull request by ysbaddaden**
The current `Crypto::BCrypt` implementation is roughly ~68% slower than the Ruby extension.

Changing the HEAP allocated Array for the stack allocated StaticArray and flattening `P` immediately reduces the difference to ~32% slower. Using raw pointers closes the gap further by going down to ~19% slower. Raw pointers are unsafe, but the algorithm has control over the pointer's range, and it never goes out of bounds. It's still not on par with the Ruby implementation (that uses C) but it's getting closer.

Running this [benchmark](https://gist.github.com/ysbaddaden/b43c515b1e08f52e7d7142e5a2f251af), generating 10 hashes at cost 12 got a ~~71%~~ 41% performance increase, down from 4.15s to 2.94s (vs 2.46s for Ruby).

refs #1309

### CODE REVIEW

File: src/crypto/blowfish.cr [hunk 1]
```diff
@@ -83,192 +82,187 @@ class Crypto::Blowfish
...
-      0x93a83531, 0x56cccd02, 0xacf08162, 0x5a75ebb5, 0x6e163697, 0x88d273cc,
-      0xde966292, 0x81b949d0, 0x4c50901b, 0x71c65614, 0xe6c6c7bd, 0x327a140a,
-      0x45e1d006, 0xc3f27b9a, 0xc9aa53fd, 0x62a80f00, 0xbb25bfe2, 0x35bdd2f6,
-      0x71126905, 0xb2040222, 0xb6cbcf7c, 0xcd769c2b, 0x53113ec0, 0x1640e3d3,
-      0x38abbd60, 0x2547adf0, 0xba38209c, 0xf746ce76, 0x77afa1c5, 0x20756060,
-      0x85cbfe4e, 0x8ae88dd8, 0x7aaaf9b0, 0x4cf9aa7e, 0x1948c25c, 0x02fb8a8c,
-      0x01c36ae4, 0xd6ebe1f9, 0x90d4f869, 0xa65cdea0, 0x3f09252d, 0xc208e69f,
-      0xb74e6132, 0xce77e25b, 0x578fdfe3, 0x3ac372e6,
-    ],
+  private S = StaticArray[
```
Comment (lribeiro) [inline src/crypto/blowfish.cr]: Why not UInt32.static_array(...) instead?
Won't it spare us all the _32?

Comment (ysbaddaden) [inline src/crypto/blowfish.cr]: Thanks. I wasn't aware of this initializer. I wish `Int[1, 2, 3]` would return a `StaticArray` instead of an `Array`...

**Comment by kostya** [comment]
replace f with, gives also 5% performance:
```crystal
  @[AlwaysInline]
  private def f(x)
    d = x.to_u8
    c = (x >> 8).to_u8
    b = (x >> 16).to_u8
    a = (x >> 24).to_u8
    ((@s.to_unsafe[a] + @s.to_unsafe[256 + b]) ^ @s.to_unsafe[512 + c]) + @s.to_unsafe[768 + d]
  end
```

**Comment by ysbaddaden** [comment]
@kostya I didn't notice an improvement (above benchmark, LLVM 3.9.1).

I got a 6% improvement, by memoizing pointers to `@s` sections, instead of repeating additions in `f` which is costly, and keeping `& 0xFF_u32` which appears to be faster than casting with `to_u8` (at least on Linux x86_64).

It's now down to 2.81s vs 2.46s (ruby).

**Comment by mverzilli** [comment]
Excellent as usual, Julien! 🙇 

Thank you so much, this is an amazing contribution 💯

**Pull request merged**